doc: document async/await* (#4660)
dfinity/motoko: doc/md/writing-motoko/async-star.md

Similar to how `await` is used to consume the result of a future, `await*` is used to produce the result of a computation (by demanding another execution of its body).

From a typing perspective, futures and computations are very similar. Where they differ is in their dynamic behavior: a future is a stateful object that holds the result of a scheduled, asynchronous task while a computation is just an inert value describing a task.

Unlike `await` on a future, `await*` on a computation does not suspend the awaiter, it just immediately executes the computation much like an ordinary function call.
This means that awaiting an `async*` value only suspends its execution (to complete asynchronously), if the body of the `async*` does a proper `await`.
The `*` on these expressions is meant to indicate that the computation may involve 0 or more ordinary `await` expressions, and thus may be interleaved with the execution of other messages.

To create an `async*` value, you can just use an `async*` expression, but more typically, you'll declare a local function that returns an `async*` type.

To compute the result of an `async*` computation, you just use an `await*`.

> 

Author: ```suggestion

Here's how we can refactor our class efficiently and without changing its meaning using computations, not futures:

```

Awaiting the same future another time will always produce the original result: the message is executed just once.

On the other hand, calling the `async*` version of `maybeLog` will do nothing unless the result is `await*`-ed, and `await*`-ing the same computation several times will repeat
the computation each time.

For another example, suppose we define a
clap function with the side-effect of printing "clap".

``` motoko name=clap
import Debug "mo:base/Debug"
func clap() { Debug.print("clap") }
```

Now, using futures, the code will clap once:

``` motoko include=clap
let future = async { clap() };
```

This remains the case, no matter how often you await `future`.
For example:

``` motoko include=clap
let future = async { clap() };
await future;
await future;
```

Using computations, on the other hand, the following definition has no effect on its own:

``` motoko include=clap
let computation = async* { clap() };
```
But, the following example will clap twice:

``` motoko include=clap
let computation = async* { clap() };
await* computation;
await* computation;
```

> 

Reviewer: Maybe it should be mentioned that both `async` and `async*` require send capabilities statically.

Author: I guess the language manual mentions that.

:::danger

> You should use `async*`/`await*` star with care. An ordinary `await` is a commit point in Motoko: all your state changes will be committed at the `await` before suspension.

Reviewer: "star" is redundant

``` motoko
actor class (Logger : actor { log : Text -> async () }) {

  var logging = true;

  func maybeLog(msg : Text) : async* () {
    if (logging) { await Logger.log(msg) };
  };

  func doStuff() : async () {
    // do stuff
    await* maybeLog("stuff");
    // do more stuff
    await* maybeLog("more stuff");
  }
}
```

One notable difference between `async` and `async*` expressions is that the former are eager, while the latter are not.
This means that calling the async version of `maybeLog` will eagerly schedule its body to run, even if the `async` result (a future) of the call is never `await`ed.
Awaiting the same future another time will always produce the original result: the message is executed just once.

On the other hand, calling the `async*` version of `maybeLog` will do nothing unless the result is `await*`-ed, and `await*`-ing the same computation several times will repeat
the computation each time.

For another example, suppose we define a
clap function with the side-effect of printing "clap".

``` motoko name=clap
import Debug "mo:base/Debug"
func clap() { Debug.print("clap") }
```

> Now, using futures, the code will clap once:

Author: ```suggestion
Now, using futures, this code will clap once:
```

``` motoko include=clap
let future = async { clap() };
```

This remains the case, no matter how often you await `future`.
For example:

``` motoko include=clap
let future = async { clap() };
await future;
await future;
```

Using computations, on the other hand, the following definition has no effect on its own:

``` motoko include=clap
let computation = async* { clap() };
```
But, the following example will clap twice:

``` motoko include=clap
let computation = async* { clap() };
await* computation;
await* computation;
```

:::danger

> You should use `async*`/`await*` star with care. An ordinary `await` is a commit point in Motoko: all your state changes will be committed at the `await` before suspension.

Author: ```suggestion
You should use `async*`/`await*` with care. An ordinary `await` is a commit point in Motoko: all your state changes will be committed at the `await` before suspension.
```

---
sidebar_position: 27
---

# Abstracting asynchronous code

Functions are an abstraction mechanism, allowing you to name a computation and re-use that computation in different locations within your code simply by invoking the name of that function. When the function takes parameters, you can tailor the computation to different call sites by providing different arguments.

Programmers often improve their code by re-factoring common patterns of code into
a single, reusable function.

In Motoko, you might want to refactor code that involves asynchronous operations such as sending messages or awaiting futures.
Motoko's type system prevents you from using an ordinary function for this because ordinary functions are not allowed to send messages nor await.

> You can, however, define a local, asynchronous function containing the asynchronous code, and then replace all occurrences of the pattern by an `await` of a call to that function.

Author: ```suggestion
You can, however, define a local, asynchronous function containing the asynchronous code, and then replace all occurrences of the pattern by a call to that function. Since these calls return futures, each call must be enclosed in an `await` to extract its future's result.
```

```
But, the following example will clap twice:

``` motoko include=clap
let computation = async* { clap() };
await* computation;
await* computation;
```

:::danger

You should use `async*`/`await*` star with care. An ordinary `await` is a commit point in Motoko: all your state changes will be committed at the `await` before suspension.
An `await*`, on the other hand, is not a commit point (since its body may not await at all, or commit at some indefinite point).
This means that traps within the awaited computation may roll back the state of the actor to the last commit point *before* the `await*`, not to the state at the `await*` itself.

:::

See the language manual for more details on the [`async*` type](../reference/language-manual#async-type-1), the [`async*` expression](../reference/language-manual#async-1) and the
[`await*` expression](../reference/language-manual#await-1).

## Mops packages for computations

- [`star`](https://mops.one/star): Used for handling asynchronous behavior and traps using `async*` functions.

<img src="https://github.com/user-attachments/assets/844ca364-4d71-42b3-aaec-4a6c3509ee2e" alt="Logo" width="150" height="150" />

> 
> 
> 
> 
> 
> 
> 
> 
> 

Author: ```suggestion
```

---
sidebar_position: 27
---

# Abstracting asynchronous code

Functions are an abstraction mechanism.
A function allows you to name a computation and re-use that computation at different locations
with your code, simply by invoking the name of the function.
When the function takes parameters, you can tailor the computation to different call sites
by providing different arguments.

Programmers often improve their code by re-factoring common patterns of code into
a single, reusable function.

> In TKO, you might want to refactor code that involves asynchronous operations such as sending messages or awaiting futures.

Author: Bugger.  Maybe my ispell corrupted it. Was probably Motoko.

> In TKO, you might want to refactor code that involves asynchronous operations such as sending messages or awaiting futures.

Author: ```suggestion
In Motoko, you might want to refactor code that involves asynchronous operations such as sending messages or awaiting futures.
```

> In TKO, you might want to refactor code that involves asynchronous operations such as sending messages or awaiting futures.

Author: Fixed

Motoko's type system prevents you from using an ordinary function for this, because ordinary functions are not allowed to send messages nor await.

> You can, however, define a local, asynchronous function containing the asynchronous code, and then replace all occurrences of the pattern by an `await` of a call to that function.

Author: No, the original is correct, but maybe there's a better phrasing? There's not such thing as an `await call` but you can  await the future returned by a call to that function.

> You can, however, define a local, asynchronous function containing the asynchronous code, and then replace all occurrences of the pattern by an `await` of a call to that function.

Author: I rephrased it a bit, adding sentence.

``` motoko
actor class (Logger : actor { log : Text -> async () }) {

  var logging = true;

  func maybeLog(log : Bool, msg : Text) : async () {
    if (log) { await Logger.log(msg) };
  };

  func doStuff() : async () {
    // do stuff
    await maybeLog(logging, "stuff");
    // do more stuff
    await maybeLog(logging, "more stuff");
  }
}
```
## Computation types

To avoid the overhead and dangers of additional awaits, Motoko offers computation types, `async* T`, that, like future types, `async T`, can abstract asynchronous tasks.

Just as an `async` expression is used to create a future (by scheduling the execution of its body), an `async*` expression is used to create a computation (by delaying the execution of its body).
And just like `await` is used to consume the result of a future, `await*` is used to produce the result of a computation (by demanding another execution of its body).

From a typing perspective, futures and computations are very similar. Where they differ is in their dynamic behaviour: a future is a stateful object that holds the result of a scheduled, asynchronous task while a computation is just an inert value describing a task.

> Unlike `await` on a future, `await*` on a computation does not suspend the awaiter, it just immediately executes the computation, much like an ordinary  function call.
> This means that awaiting an `async*` value only suspends its execution (to complete asynchronously), if the body of the `async*` does a proper `await`.
> The `*` on these expressions is meant to indicate that the computation may involve 0 or more ordinary `await` expressions,
> and thus may be interleaved with the execution of other messages.
> 
> To create an `async*` value you can just use an `async*` expression, but, more typically, you'll declare a local function that returns an `async*` type.

Reviewer 2: ```suggestion
Unlike `await` on a future, `await*` on a computation does not suspend the awaiter, it just immediately executes the computation much like an ordinary function call.
This means that awaiting an `async*` value only suspends its execution (to complete asynchronously), if the body of the `async*` does a proper `await`.
The `*` on these expressions is meant to indicate that the computation may involve 0 or more ordinary `await` expressions, and thus may be interleaved with the execution of other messages.

To create an `async*` value, you can just use an `async*` expression, but more typically, you'll declare a local function that returns an `async*` type.
```

To compute the result of an `async*` computation, you just use an `await*`.

``` motoko
actor class (Logger : actor { log : Text -> async () }) {

  var logging = true;

  func maybeLog(msg : Text) : async* () {
    if (logging) { await Logger.log(msg) };
  };

  func doStuff() : async () {
    // do stuff
    await* maybeLog("stuff");
    // do more stuff
    await* maybeLog("more stuff");
  }
}
```

One notable difference between `async` and `async*` expressions is that the former are eager, while the latter are not.
This means that calling the async version of `maybeLog` will eagerly schedule its body to run, even if the `async` result (a future) of the call is never `await`ed.
Awaiting the same future another time will always produce the original result: the message is executed just once.

On the other hand, calling the `async*` version of `maybeLog` will do nothing unless the result is `await*`-ed, and `await*`-ing the same computation several times will repeat
the computation each time.

For another example, suppose we define a
clap function with the side-effect of printing "clap".

``` motoko name=clap
import Debug "mo:base/Debug"
func clap() { Debug.print("clap") }
```

> Now, using futures, the code:

Reviewer 2: ```suggestion
Now, using futures, the code will clap once:
```

``` motoko include=clap
let future = async { clap() };
```

> claps once. That remains the case, no matter how often you await `future`.

Reviewer 2: ```suggestion
This remains the case, no matter how often you await `future`.
```

> For example,

Reviewer 2: ```suggestion
For example:
```

``` motoko include=clap
let future = async { clap() };
await future;
await future;
```

> only claps once.

Reviewer 2: ```suggestion

```

> Using computations, on the other hand, the definition:

Reviewer 2: ```suggestion
Using computations, on the other hand, the following definition has no effect on its own:
```

``` motoko include=clap
let computation = async* { clap() };
```

> has no effect on its own, but

Reviewer 2: ```suggestion
But, the following example will clap twice:
```

``` motoko include=clap
let computation = async* { clap() };
await* computation;
await* computation;
```

> claps twice.

Reviewer 2: ```suggestion

```

:::danger

> You should use `async*`/`await*` star with care. An ordinary `await` is a commit point in Motoko: all your state changes will be committed at the `await`, before suspension.

Reviewer 2: ```suggestion
You should use `async*`/`await*` star with care. An ordinary `await` is a commit point in Motoko: all your state changes will be committed at the `await` before suspension.
```

An `await*`, on the other hand, is not a commit point (since its body may not await at all, or commit at some indefinite point).

> This means that traps within the awaited computation,  may roll back the state of the actor to the last commit point *before* the `await*`, not to the state at the `await*` itself.

Reviewer 2: ```suggestion
This means that traps within the awaited computation may roll back the state of the actor to the last commit point *before* the `await*`, not to the state at the `await*` itself.
```
